ctlra/devices/3dconnexion.c

```diff
@@ -0,0 +1,291 @@
+struct spacemouse_t {
+	struct ctlra_dev_t base;
+	int16_t hw_values[CONTROLS_SIZE];
+	uint8_t buttons[BUTTONS_SIZE];
+};
+
+static const char *
+spacemouse_control_get_name(enum ctlra_event_type_t type,
+			    uint32_t control)
+{
+	switch(type) {
+	case CTLRA_EVENT_SLIDER:
+		if(control >= CONTROL_NAMES_SLIDERS_SIZE)
+			return 0;
+		return spacemouse_names_sliders[control];
+	case CTLRA_EVENT_BUTTON:
+		if(control >= CONTROL_NAMES_BUTTONS_SIZE)
+			return 0;
+		return spacemouse_names_buttons[control];
+	default:
+		break;
+	}
+	return 0;
+}
+
+static uint32_t
+spacemouse_poll(struct ctlra_dev_t *base)
+{
+	struct spacemouse_t *dev = (struct spacemouse_t *)base;
+#define BUF_SIZE 32
+	uint8_t buf[BUF_SIZE];
+	ctlra_dev_impl_usb_interrupt_read(base, USB_HANDLE_IDX,
+					  USB_ENDPOINT_READ, buf,
+					  BUF_SIZE);
+	return 0;
+}
+
+void
+spacemouse_usb_read_cb(struct ctlra_dev_t *base, uint32_t endpoint,
+		       uint8_t *data, uint32_t size)
+{
+	struct spacemouse_t *dev = (struct spacemouse_t *)base;
+	uint8_t *buf = data;
+
+	switch(size) {
+	case 7:
+		for(int i = 0; i < BUTTONS_SIZE; i++) {
+			int p = buf[buttons[i].offset] & buttons[i].mask;
+			if(p == dev->buttons[i])
+				continue;
+			dev->buttons[i] = p;
+
+			struct ctlra_event_t event = {
+				.type = CTLRA_EVENT_BUTTON,
+				.button  = {
+					.id = i,
+					.pressed = p
+				},
+			};
+			struct ctlra_event_t *e = {&event};
+			dev->base.event_func(&dev->base, 1, &e,
+					     dev->base.event_func_userdata);
+		}
+		break;
+	case 13:
+		for(int i = 0; i < CONTROLS_SIZE; i++) {
+			int off = i * 2;
+			int16_t v = buf[off + 1] | (buf[off + 2] << 8);
+			if(v == dev->hw_values[i])
+				continue;
+
+			dev->hw_values[i] = v;
+			int neg = v < 0;
+			if(neg)
+				v *= -1;
+
+			struct ctlra_event_t event = {
+				.type = CTLRA_EVENT_SLIDER,
+				.slider = {
+					.id = (off - neg) + 1,
+					.value = v / 350.f},
+			};
+			struct ctlra_event_t *e = {&event};
+			dev->base.event_func(&dev->base, 1, &e,
+					     dev->base.event_func_userdata);
+		}
+		break;
+	default: break;
+	}
+}
+
+static void
+spacemouse_light_set(struct ctlra_dev_t *base,
+				 uint32_t light_id,
+				 uint32_t light_status)
+{
+	struct spacemouse_t *dev = (struct spacemouse_t *)base;
+	(void)dev;
+	(void)light_id;
+	(void)light_status;
+}
+
+static void
+spacemouse_light_flush(struct ctlra_dev_t *base, uint32_t force)
+{
+	struct spacemouse_t *dev = (struct spacemouse_t *)base;
+	(void)dev;
+	(void)force;
+}
+
+static int32_t
+spacemouse_disconnect(struct ctlra_dev_t *base)
+{
+	struct spacemouse_t *dev = (struct spacemouse_t *)base;
+	ctlra_dev_impl_usb_close(base);
+	free(dev);
+	return 0;
+}
+
+struct ctlra_dev_info_t ctlra_spacemouse_info;
+
+struct ctlra_dev_t *
+ctlra_spacemouse_connect(ctlra_event_func event_func, void *userdata,
+			    void *future)
+{
+	(void)future;
+	struct spacemouse_t *dev = calloc(1, sizeof(struct spacemouse_t));
+	if(!dev)
+		goto fail;
```

Comment:
This is the only goto fail, but there are 3 instances of the code of fail in this function.

Reply:
Fixed, thanks for review!